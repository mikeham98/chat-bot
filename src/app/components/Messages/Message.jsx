import React from 'react';
import styles from '../../../themes/messages.scss';

const Message = ({body, dateTime}) => {
    return (
        <div className={styles.message} data-tip={dateTime}>
            {body}
        </div>
    );
};

export default React.memo(Message);