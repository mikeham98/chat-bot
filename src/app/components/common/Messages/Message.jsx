import React from 'react';
import styles from '../../../../themes/messages.scss';

const Message = ({body}) => {
    return (
        <div className={styles.message}>
            {body}
        </div>
    );
};

export default React.memo(Message);