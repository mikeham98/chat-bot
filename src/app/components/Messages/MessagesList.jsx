import React from 'react';
import moment from 'moment';
import classnames from 'classnames';
import styles from '../../../themes/messages.scss';
import isPopulatedArray from "../../util/isPopulatedArray";
import Message from "./Message";
import {left, right} from "../../config/messages.config";
import formatDateTime from "../../util/formatDateTime";

export default class MessagesList extends React.PureComponent {
    constructor(props) {
        super(props);
        this.state = {
            clickedMessageId: null
        };
        this.messageContainerRef = this.messageContainerRef.bind(this);
    }

    componentDidMount() {
        this.scrollToBottom();
    }

    componentDidUpdate(prevProps) {
        if ((this.props.replying && !prevProps.replying) || (this.props.messages.length > prevProps.messages.length)) {
            this.scrollToBottom();
        }
    }

    scrollToBottom() {
        this.messageContainer.scrollTop = this.messageContainer.scrollHeight;
    }

    returnMessageClassName(position) {
        const classNames = [styles.messageWrapper];
        if (position === right) {
            classNames.push(styles.rightMessage);
        } else {
            classNames.push(styles.leftMessage);
        }
        return classnames(classNames);
    }

    returnMessages() {
        const {messages, currentUserId} = this.props;
        if (isPopulatedArray(messages)) {
            return messages.map(message => {
                const position = currentUserId === message.createdBy.userId ? right : left;
                // showDateTime is implemented so that only one message at a time can show the date/time
                const showDateTime = message.id === this.state.clickedMessageId;
                return (
                    <div key={message.id} onClick={() => this.setState({clickedMessageId: showDateTime ? null : message.id})} className={this.returnMessageClassName(position)}>
                        <Message
                            showDateTime={showDateTime}
                            dateTime={formatDateTime(message.createdAt)}
                            body={message.body}
                        />
                    </div>
                );
            })
        }
    }

    messageContainerRef(ref) {
        this.messageContainer = ref;
    }

    render() {
        const {replying} = this.props;
        return (
            <div
                ref={this.messageContainerRef}
                className={styles.messageContainer}
            >
                {this.returnMessages()}
                {replying && <span className={styles.replyInProgress}>ChatBot is replying...</span>}
            </div>
        )
    }
}