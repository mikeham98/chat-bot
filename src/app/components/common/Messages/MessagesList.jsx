import React from 'react';
import moment from 'moment';
import classnames from 'classnames';
import styles from '../../../../themes/messages.scss';
import isPopulatedArray from "../../../util/isPopulatedArray";
import Message from "./Message";
import {left, right} from "../../../config/messages.config";

export default class MessagesList extends React.PureComponent {
    componentDidMount() {
        this.messageContainer.scrollTop = this.messageContainer.scrollHeight;
    }

    returnMessageClassName(position) {
        const classNames = [styles.messageWrapper];
        if(position === right) {
            classNames.push(styles.rightMessage);
        }else {
            classNames.push(styles.leftMessage);
        }
        return classnames(classNames);
    }

    returnMessages() {
        const {messages, currentUserId} = this.props;
        if (isPopulatedArray(messages)) {
            return messages.map(message => {
                const position = currentUserId === message.createdBy.userId ? right : left;
                return (
                    <div className={this.returnMessageClassName(position)}>
                        <Message
                            dateTime={moment(message.createdAt).fromNow()}
                            body={message.body}
                        />
                    </div>
                );
            })
        }
    }

    render() {
        return (
            <div
                ref={(ref) => this.messageContainer = ref}
                className={styles.messageContainer}
            >
                {this.returnMessages()}
            </div>
        )
    }
}