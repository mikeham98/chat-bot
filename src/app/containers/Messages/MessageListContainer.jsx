import React from 'react';
import {connect} from 'react-redux';
import MessagesList from "../../components/Messages/MessagesList";
import {getConversationMessages} from "../../actions/conversation/index.actions";

export class MessageListContainer extends React.Component{
    componentDidMount() {
        this.props.getConversationMessages();
    }

    render() {
        const {messages, replying, currentUserId} = this.props;
        return (
            <MessagesList
                messages={messages}
                replying={replying}
                currentUserId={currentUserId}
            />
        )
    }
}

const mapStateToProps = (state, ownProps) => {
    return {
        replying: state.conversation.replying,
        messages: state.conversation.messages
    }
};

export default connect(mapStateToProps, {getConversationMessages})(MessageListContainer);