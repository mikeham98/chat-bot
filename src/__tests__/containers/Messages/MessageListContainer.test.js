import React from 'react';
import {shallow} from 'enzyme';
import {MessageListContainer} from '../../../app/containers/Messages/MessageListContainer';

describe('MessageListContainer', () => {

    let shallowedWrapper;
    let props = {};
    const wrapper = () => {
        if (!shallowedWrapper) {
            shallowedWrapper = shallow(<MessageListContainer {...props}/>);
        }
        return shallowedWrapper;
    };

    beforeEach(() => {
        props = {
            getMessages: jest.fn(),
            messages: [1, 2, 3, 4, 5],
            replying: true,
            currentUserId: 1
        };
        shallowedWrapper = undefined;
    });

    describe('render', () => {
        it('should return MessageList component', () => {
            expect(wrapper()).toMatchSnapshot();
        });
    });

    describe('functionality', () => {
        it('should call getMessages on mount', () => {
            expect(props.getMessages).not.toHaveBeenCalled();
            wrapper();
            expect(props.getMessages).toHaveBeenCalledTimes(1);
        });
    });
});