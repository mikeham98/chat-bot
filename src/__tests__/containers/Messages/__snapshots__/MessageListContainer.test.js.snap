// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageListContainer render should return MessageList component 1`] = `
<MessagesList
  currentUserId={1}
  messages={
    Array [
      1,
      2,
      3,
      4,
      5,
    ]
  }
  replying={true}
/>
`;
